clog/loggers.py

```diff
@@ -295,27 +304,57 @@ def log_line(self, stream, line):
             self.metrics.monk_exception()
 
     def _log_line_no_size_limit(self, stream, line):
-        if time.time() < self.last_timeout + self.timeout_backoff_s:
+        now = time.time()
+        if now - self.last_disconnect < self.timeout_backoff_s:
+            if self.buffering:
+                self._add_to_buffer(stream, line)
             return
+        elif self.buffering:
+            self.buffering = False
+            self._flush_buffer()
+            self.report_status(False, 'Flushed buffer ({} left)'.format(len(self.buffer)))
+
         with self.metrics.sampled_request():
             try:
                 self.producer.send_messages(
                     self.stream_prefix + stream,
                     [line],
                     None
                 )
-            except socket.timeout as e:
-                self.report_status(True, 'Monk took too long to respond')
-                self.last_timeout = time.time()
-                self.metrics.monk_timeout()
             except Exception as e:
-                self.report_status(
-                    True,
-                    'Exception while sending to monk: %s' % str(e)
-                )
-                self.metrics.monk_exception()
+                if isinstance(e, socket.timeout):
+                    self.report_status(True, 'Monk took too long to respond')
+                    self.metrics.monk_timeout()
+                else:
+                    self.report_status(True, 'Exception while sending to monk: %s' % str(e))
+                    self.metrics.monk_exception()
+
+                self.last_disconnect = now
+                if self.use_buffer:
+                    self.buffering = True
+                    self.report_status(False, 'Start buffering')
+                    self.metrics.monk_buffer()
```

Comment:
The "monk_buffer" metric is intended to count the number of lines that are being added to the buffer, right? Might be worth moving it inside the `_add_to_buffer`. At the moment we are not counting the ones added at the beginning of this method.

Reply:
I intended it to count the number of times clog starts buffering. Sending a metric datapoint each time a line is buffered would add some time, we could use the same mechanism as other metrics and send them when we sample. No real preference